Refactor delete methods in the database

pstaabp opened the pull request:
This refactors all of the deleteXXX methods in the database layer.  

- Previously `deleteXXX` would return the deleted object.  
- At the db layer, nothing is returned. 
- At the mojolicious layer, a message saying XXX was deleted successfully.

In addition
- all perl tests now delete objects then check that they are successfully deleted. 
- the client-side store tests also delete the object then checks the database for successful deletion. 

Note: The UI never used a deleted object so no UI changes should occur.

pstaabp committed:
WIP: change deleteXXX in ResultSet classes.

pstaabp committed:
Change the deleteXXX mojolicious tests.

pstaabp committed:
WIP: deleting from db--update store tests

WIP: cleanup the tests


WIP: deleting from db--update store tests

pstaabp committed:
WIP: send message from mojolicious layer and more cleanup.

pstaabp committed:
FIX: update permission

pstaabp committed:
FIX: regression after rebase.

drdrew42 commented on lib/DB/Schema/ResultSet/ProblemPool.pm:
```diff
@@ -374,12 +435,8 @@ remove a Problem out of a ProblemPool in a course
-
-	my $prob2 = $prob->delete;
-	return $prob2 if $args{as_result_set};
-	return { $prob2->get_inflated_columns };
+	$self->getPoolProblem(info => $args{info}, as_result_set => 1)->delete;
+	return;
```
Not part of this PR, but why is this called `remove` instead of `delete`?

pstaabp replied:
My thought is that we are removing a problem from a pool, not deleting the problem.  Happy to switch to delete though.

drdrew42 commented on src/stores/problem_sets.ts:
```diff
@@ -268,16 +272,21 @@ export const useProblemSetStore = defineStore('problem_sets', {
-			user_set.set(deleted_user_set.toObject());
-			return user_set;
+			if (response.status === 200) {
+				const index = this.db_user_sets.findIndex((set) => set.user_set_id === user_set.user_set_id);
+				if (index < 0) {
+					logger.error('[user store/deleteUserSet]: the user set was not found in the store');
```
Shouldn't we be checking that the user set is present before making the API call? This is an error that 'should never happen', but it doesn't make sense to check afterwards.

Also, I worry about having removed the Exceptions on the DB side for attempting to delete objects that don't exist... that should be an error still, and the response wouldn't be 200 OK.

pstaabp replied:
Will switch the order of these.  About the exception, see my comment below.

drdrew42 commented on src/stores/set_problems.ts:
```diff
@@ -166,16 +166,18 @@ export const useSetProblemStore = defineStore('set_problems', {
-			const index = this.set_problems.findIndex(prob => prob.set_problem_id === deleted_problem.set_problem_id);
-			this.set_problems.splice(index, 1);
-			return deleted_problem;
+			const index = this.set_problems.findIndex(prob => prob.set_problem_id === problem.set_problem_id);
+			if (index < 0) {
+				logger.error('[stores/set_problems/deleteSetProblem]: the set problem was not found in the store');
```
Check before API

drdrew42 commented on src/stores/set_problems.ts:
```diff
@@ -250,26 +252,25 @@ export const useSetProblemStore = defineStore('set_problems', {
-				.findIndex(user_problem => user_problem.user_problem_id === db_deleted_problem.user_problem_id);
-			if (index >= 0) {
-				this.db_user_problems.splice(index, 1);
+				.findIndex(user_problem => user_problem.user_problem_id === user_problem.user_problem_id);
+			if (index < 0) {
+				logger.error('[stores/set_problems/deleteUserProblem]: the set problem was not found in the store');
```
Check before API

drdrew42 commented on src/stores/users.ts:
```diff
@@ -155,15 +155,20 @@ export const useUserStore = defineStore('user', {
 				const index = this.users.findIndex((u) => u.user_id === user.user_id);
-				// splice is used so vue3 reacts to changes.
-				this.users.splice(index, 1);
-				return new User(response.data as ParseableUser);
+				if (index < 0) {
+					logger.error('[user store/deleteUser]: the user was not found in the store');
```
Check before API

drdrew42 commented on src/stores/users.ts:
```diff
@@ -261,16 +266,16 @@ export const useUserStore = defineStore('user', {
-				this.db_course_users.splice(index, 1);
-				const deleted_course_user = new DBCourseUser(response.data as ParseableCourseUser);
-				const user = this.users.find(u => u.user_id === deleted_course_user.user_id);
-				return new CourseUser(Object.assign({}, user?.toObject(), deleted_course_user.toObject()));
+				if (index < 0) {
+					logger.error('[user store/deleteCourseUser]: the user was not found in the store');
```
Check before API

drdrew42 reviewed (commented):
I'm concerned about lack of Exceptions on delete calls. Roughly speaking, IMO, the approach should be:
1. Fetch the object to be deleted
2. If the object doesn't exist, throw NotFoundException
3. Delete the object

And on the UI side:
1. Check that the object exists in the store
2. If the object doesn't exist, display error message and halt
3. Make API call
4. Handle API response (update store or handle exception message)

pstaabp commented:
There are still exceptions thrown.  For example, that you mention above, on `ProblemPool`, the `getProblemPool` will throw an exception if it doesn't exist. This is how all of the deleteXXX methods work as well.

pstaabp committed:
FIX: refactor deleteXXX functions to check for existence before making API call.

FIX: refactor deleteXXX functions to check for existence before making API call.
